File: Homework/Hoang_Nguyen/lesson7/144_binaryTreePreorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
  const result = [];
  dfs(root);
  return result;

  function dfs(node) {
    if (!node) return;
    result.push(node.val);
    dfs(node.left);
    dfs(node.right);
  }
};

// iterative solution
var preorderTraversal = function (root) {
  const result = [];
  const stack = [];
  stack.push(root);
  while (stack.length) {
    const node = stack.pop();
    if (!node) continue;
    result.push(node.val);
    stack.push(node.right);
    stack.push(node.left);
  }
  return result;
};

File: Homework/Ngan_Luu/144-binary-tree-preorder-traversal.py
144. https://leetcode.com/problems/binary-tree-preorder-traversal
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res=[]
        def preOrder(node):
            if not(node):
                return
            res.append(node.val)
            
            preOrder(node.left)
                
            preOrder(node.right)
        preOrder(root)
        return res
        